### PR TITLE
Add logic to load datatype info hash table on demand

### PR DESCRIPTION
**Pull request by rishabhtanwar29**
CAST to sql_variant datatype was failing during restore with the
cashe lookup failed error, which is because we are using `type_infos`
array to get oid of certain datatypes for sql_variant input/output
and comparison functions but we only initialize `type_infos` array
once during __pg_init. But during pg_upgrade since __pg_init gets
run before creating any datatypes so `type_infos` array has all the
oids as `0` (invalidOid).

This commit fixes this by making hash table initialization on demand.
Introduced a helper function `get_tsql_type_info` which makes
sure to initialize hash table before accessing it.
Added test in upgrade format so that the scenario gets tested in
both normal and upgrade tests.

Task: BABEL-3404
Signed-off-by: Rishabh Tanwar <ritanwar@amazon.com>

### Description

[Describe what this change achieves]
 
### Issues Resolved

[List any issues this PR will resolve]

### Check List
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: test/JDBC/input/datatypes/TestSQLVariant-vu-prepare.sql [hunk 1]
```diff
@@ -261,3 +261,11 @@ Insert into testsqlvariant_sourceTable22 values (cast ('2016-10-23 12:45:37.123'
 go
 Insert into testsqlvariant_sourceTable22 values (cast ('2016-10-23 12:45:37.12345' as datetime2(5)), cast ('2016-10-23 12:45:37.123456' as datetime2(6)));
 go
+
+-- Test sql_variant cast (Test for BABEL-3404)
+CREATE VIEW testsqlvariant_view1 AS SELECT CAST('abc' AS sql_variant) AS col;
```
Comment (surendravishnoi) [inline test/JDBC/input/datatypes/TestSQLVariant-vu-prepare.sql]: object name should be filename_xxx

File: contrib/babelfishpg_common/src/typecode.c [hunk 2]
```diff
@@ -114,11 +118,29 @@ init_tcode_trans_tab(PG_FUNCTION_ARGS)
             entry = hash_search(ht_oid2typecode, &type_infos[i].oid, HASH_ENTER, NULL);
             entry->persist_id = i;
         }
+        else
+        {
+            /* type is not loaded. wait for next scan */
+			inited_ht_oid2typecode = false;
```
Comment (surendravishnoi) [inline contrib/babelfishpg_common/src/typecode.c]: indentation

Comment (rishabhtanwar29) [reply contrib/babelfishpg_common/src/typecode.c]: Fixed!

File: test/JDBC/input/datatypes/TestSQLVariant-vu-prepare.sql (hunk 1 above)
Comment (rishabhtanwar29) [reply test/JDBC/input/datatypes/TestSQLVariant-vu-prepare.sql]: Are you suggesting that we should also include `-vu-prepare` part in the object name?

Comment (surendravishnoi) [reply test/JDBC/input/datatypes/TestSQLVariant-vu-prepare.sql]: yes, whatever is in the file name should be the prefix for object name. We want to keep the rule simple, always use filename as prefix.

Comment (rishabhtanwar29) [reply test/JDBC/input/datatypes/TestSQLVariant-vu-prepare.sql]: Updated the object name!

**Review by surendravishnoi — DISMISSED** [review]

File: test/JDBC/expected/TestSQLVariant-vu-verify.out [hunk 3]
```diff
@@ -267,3 +267,22 @@ sql_variant#!#sql_variant
...
+
+Select * from testsqlvariant_vu_prepare_view1;
+go
+~~START~~
+sql_variant
+abc
+~~END~~
+
+
+Select * from testsqlvariant_vu_prepare_view2;
```
Comment (2jungkook) [inline test/JDBC/expected/TestSQLVariant-vu-verify.out:279]: I'd prefer to put `order by` if multiple rows are expected.

**Review by 2jungkook — APPROVED** [review]

File: test/JDBC/expected/TestSQLVariant-vu-verify.out (hunk 3 above)
Comment (rishabhtanwar29) [reply test/JDBC/expected/TestSQLVariant-vu-verify.out:279]: Currently the output seems consistent so I think we can skip the clause.

**Review by surendravishnoi — APPROVED** [review]

**Pull request merged**